File: python/testing/test_interactor.py
import pytest

import f3d


def print_fn(args: list[str]):
    print(args)


def doc_fn():
    return "dummyDoc", "dummyValue"


def compl_fn(args: list[str]):
    return ["compl"]


def test_command(capfd: pytest.CaptureFixture[str]):
    engine = f3d.Engine.create(True)
    inter = engine.interactor

    # Check default commands can be removed
    actions = inter.get_command_actions()
    for action in actions:
        inter.remove_command(action)
    assert len(inter.get_command_actions()) == 0

    # Check a command can be triggered
    inter.add_command("my_cmd", print_fn)
    inter.trigger_command("my_cmd arg1 arg2")
    inter.remove_command("my_cmd")
    out, _err = capfd.readouterr()
    assert out == "['arg1', 'arg2']\n"

    # Check a command can be triggered with #
    inter.add_command("my_cmd", print_fn)
    inter.trigger_command("my_cmd arg1 arg2 #000", False)
    inter.remove_command("my_cmd")
    out, _err = capfd.readouterr()
    assert out == "['arg1', 'arg2', '#000']\n"

    # Smoke test
    inter.init_commands()
    inter.add_command("my_cmd2", print_fn, ("my_cmd2", "doc"), compl_fn)


def test_command_completion(capfd: pytest.CaptureFixture[str]):
    engine = f3d.Engine.create(True)
    inter = engine.interactor

    # Check default commands can be removed
    actions = inter.get_command_actions()
    for action in actions:
        inter.remove_command(action)
    assert len(inter.get_command_actions()) == 0

    # Smoke test
    inter.init_commands()
    inter.add_command("cmd", print_fn, ("cmd", "doc"), compl_fn)

    engine.window.render()

    inter.trigger_keyboard_key(inter.InputAction.PRESS, "Escape")
    inter.trigger_keyboard_key(inter.InputAction.RELEASE, "Escape")
    for c in "cmd ":
        inter.trigger_text_character(ord(c))
    inter.trigger_keyboard_key(inter.InputAction.PRESS, "Tab")
    inter.trigger_keyboard_key(inter.InputAction.RELEASE, "Tab")


def test_binding():
    engine = f3d.Engine.create(True)
    inter = engine.interactor

    # Check default interactions can be removed
    groups = inter.get_bind_groups()
    for group in groups:
        binds = inter.get_binds_for_group(group)
        for bind in binds:
            inter.remove_binding(bind)
    assert len(inter.get_bind_groups()) == 0
    assert len(inter.get_binds()) == 0

    # Smoke test
    inter.add_binding(
        f3d.InteractionBind(f3d.InteractionBind.ModifierKeys.ANY, "P"),
        "dummy command",
        "DummyGroup",
        doc_fn,
    )
    inter.add_binding(
        f3d.InteractionBind(f3d.InteractionBind.ModifierKeys.NONE, "P"),
        "dummy command",
        "DummyGroup",
        doc_fn,
        f3d.Interactor.BindingType.CYCLIC,
    )
    inter.add_binding(
        f3d.InteractionBind(f3d.InteractionBind.ModifierKeys.CTRL, "P"),
        "dummy command",
        "DummyGroup",
        doc_fn,
        f3d.Interactor.BindingType.NUMERICAL,
    )
    inter.add_binding(
        f3d.InteractionBind(f3d.InteractionBind.ModifierKeys.SHIFT, "P"),
        "dummy command",
        "DummyGroup",
        doc_fn,
        f3d.Interactor.BindingType.TOGGLE,
    )
    inter.add_binding(
        f3d.InteractionBind(f3d.InteractionBind.ModifierKeys.CTRL_SHIFT, "P"),
        ["dummy command", "dummy command"],
        "DummyGroup",
        doc_fn,
        f3d.Interactor.BindingType.OTHER,
    )
    assert len(inter.get_bind_groups()) == 1
    assert len(inter.get_binds()) == 5
    assert (
        inter.get_binding_documentation(
            f3d.InteractionBind(f3d.InteractionBind.ModifierKeys.CTRL, "P")
        )
        == doc_fn()
    )
    assert (
        inter.get_binding_type(
            f3d.InteractionBind(f3d.InteractionBind.ModifierKeys.CTRL, "P")
        )
        == f3d.Interactor.BindingType.NUMERICAL
    )

    inter.init_bindings()


def test_trigger_key(capfd: pytest.CaptureFixture[str]):
    engine = f3d.Engine.create(True)

    engine.window.render()

    engine.interactor.trigger_keyboard_key(f3d.Interactor.InputAction.PRESS, "E")
    engine.interactor.trigger_keyboard_key(f3d.Interactor.InputAction.RELEASE, "E")
    assert engine.options["render.show_edges"] is True

    engine.interactor.trigger_mod_update(f3d.Interactor.InputModifier.SHIFT)
    engine.interactor.trigger_keyboard_key(f3d.Interactor.InputAction.PRESS, "L")
    engine.interactor.trigger_keyboard_key(f3d.Interactor.InputAction.RELEASE, "L")
    assert engine.options["render.light.intensity"] == 0.98

    engine.interactor.trigger_mouse_button(
        f3d.Interactor.InputAction.PRESS, f3d.Interactor.MouseButton.LEFT
    )
    engine.interactor.trigger_mouse_position(100, 100)
    engine.interactor.trigger_mouse_wheel(f3d.Interactor.WheelDirection.FORWARD)
    engine.interactor.trigger_text_character(0)

File: webassembly/testing/test_interactor.js
import utils from "./utils.js";

const settings = {
  runBefore: (Module) => {
    // does nothing but called for coverage
    Module.engineInstance.getScene().clear();

    const options = Module.engineInstance.getOptions();

    // background must be set to black for proper blending with transparent canvas
    options.setAsString("render.background.color", "#000000");

    // display widgets
    options.toggle("ui.axis");
    options.toggle("render.grid.enable");
  },

  runAfter: (Module) => {
    const interactor = Module.engineInstance.getInteractor();

    // commands
    interactor.initCommands();

    let callbackCalled = false;

    interactor.addCommand("foo", (args) => {
      utils.assert(args[0] === "bar");
      callbackCalled = true;
    });

    utils.assert(
      interactor.getCommandActions().includes("foo"),
      "command not added",
    );

    interactor.triggerCommand("foo bar baz", true);

    utils.assert(callbackCalled, "command callback not called");

    interactor.removeCommand("foo");
    utils.assert(
      !interactor.getCommandActions().includes("foo"),
      "command not removed",
    );

    // animations
    utils.assert(
      !interactor.isPlayingAnimation(),
      "animation should not be playing",
    );

    interactor.startAnimation(Module.InteractorAnimationDirection.FORWARD);
    utils.assert(
      interactor.isPlayingAnimation() &&
        interactor.getAnimationDirection() ==
          Module.InteractorAnimationDirection.FORWARD,
      "animation should be playing forward",
    );
    interactor.stopAnimation();
    utils.assert(
      !interactor.isPlayingAnimation(),
      "animation should not be playing",
    );

    interactor.startAnimation(Module.InteractorAnimationDirection.BACKWARD);
    utils.assert(
      interactor.isPlayingAnimation() &&
        interactor.getAnimationDirection() ==
          Module.InteractorAnimationDirection.BACKWARD,
      "animation should be playing backward",
    );
    interactor.stopAnimation();

    interactor.toggleAnimation(Module.InteractorAnimationDirection.FORWARD);
    utils.assert(
      interactor.isPlayingAnimation() &&
        interactor.getAnimationDirection() ==
          Module.InteractorAnimationDirection.FORWARD,
      "animation should be playing forward",
    );
    interactor.stopAnimation();

    interactor.toggleAnimation(Module.InteractorAnimationDirection.BACKWARD);
    utils.assert(
      interactor.isPlayingAnimation() &&
        interactor.getAnimationDirection() ==
          Module.InteractorAnimationDirection.BACKWARD,
      "animation should be playing backward",
    );
    interactor.stopAnimation();

    // bindings
    interactor.initBindings();

    const initialBinds = interactor.getBinds();
    utils.assert(
      Array.isArray(initialBinds) && initialBinds.length > 0,
      "bindings should be returned",
    );

    const bind = new Module.InteractionBind();
    bind.mod = Module.InteractionBindModifierKeys.NONE;
    bind.inter = "R";

    interactor.addBinding(
      bind,
      ["set render.grid.reflection .8"],
      "test-group",
      () => ["foo", "bar"],
      Module.InteractorBindingType.OTHER,
      false,
    );

    const updatedBinds = interactor.getBinds();
    utils.assert(
      updatedBinds.length === initialBinds.length + 1,
      "new binding should be registered",
    );

    const bindGroups = interactor.getBindGroups();
    utils.assert(
      bindGroups.includes("test-group"),
      "new bind group should be registered",
    );

    const docs = interactor.getBindingDocumentation(bind);
    utils.assert(
      Array.isArray(docs) && docs[0] === "foo" && docs[1] === "bar",
      "binding documentation should be returned",
    );

    const bindingType = interactor.getBindingType(bind);
    utils.assert(
      bindingType === Module.InteractorBindingType.OTHER,
      "binding type should round-trip",
    );

    // start recording
    interactor.recordInteraction("/test_interaction.txt");

    // input events coverage, just check we do not throw
    // all these calls have no effect on the resulting image
    let inputEventError = false;
    try {
      interactor.triggerModUpdate(Module.InteractionBindModifierKeys.NONE);
      interactor.triggerMousePosition(0, 0);
      interactor.triggerMouseButton(
        Module.InteractorInputAction.PRESS,
        Module.InteractorMouseButton.LEFT,
      );
      interactor.triggerMouseButton(
        Module.InteractorInputAction.RELEASE,
        Module.InteractorMouseButton.LEFT,
      );
      interactor.triggerTextCharacter(0x42);
    } catch {
      inputEventError = true;
    }
    utils.assert(!inputEventError, "input event bindings should not throw");

    // check file content
    const interactionFileContent = Module.FS.readFile("/test_interaction.txt", {
      encoding: "utf8",
    });
    utils.assert(
      interactionFileContent.includes("MouseMoveEvent 0 299 0 0 0 0 0"),
      "interaction file should contain the expected command",
    );

    // play it (no effect on the resulting image, just check we do not throw)
    interactor.playInteraction("/test_interaction.txt", 0.1);

    // zoom out
    interactor.triggerMouseWheel(Module.InteractorWheelDirection.BACKWARD);

    // trigger "set render.grid.reflection .8"
    interactor.triggerKeyboardKey(Module.InteractorInputAction.PRESS, "R");
    interactor.triggerKeyboardKey(Module.InteractorInputAction.RELEASE, "R");

    let bindingRemoved = false;
    try {
      interactor.removeBinding(bind);
    } catch {
      bindingRemoved = true;
    }
    utils.assert(!bindingRemoved, "binding should be removable");

    // notifications
    let notifCount = 0;

    interactor.setNotificationCallback((desc, value, bind, duration) => {
      notifCount++;
      return true;
    });

    interactor.triggerNotification("Test notification", "value", 1.0);
    utils.assert(notifCount === 1, "notification callback not called");

    let eventLoopState = null;
    interactor.setEventLoopUserCallback((state) => {
      eventLoopState = state;
    });
    interactor.triggerEventLoop(0.1);
    utils.assert(
      eventLoopState && typeof eventLoopState.animationTime === "number",
      "event loop state should expose animation time",
    );

    // only for coverage, do not test the actual feature yet
    interactor.disableCameraMovement();
    interactor.enableCameraMovement();
    interactor.requestRender();
    interactor.requestStop();
    interactor.stop();
  },
};

utils.runRenderTest(settings, {
  data: "f3d.glb",
  baseline: "TestWasmInteraction.png",
});
